liberty-maven-plugin/src/main/java/net/wasdev/wlp/maven/plugins/server/DevMojo.java

```diff
@@ -0,0 +1,882 @@
+            util.runTestThread(false, executor, -1, false, false);
+        } else {
+            // else watch for keypresses immediately
+            util.runHotkeyReaderThread(executor);
+        }
+
+        // pom.xml
+        File pom = project.getFile();
+        
+        util.watchFiles(pom, outputDirectory, testOutputDirectory, executor, artifactPaths, configFile);
+    }
+
+    private void addArtifacts(org.eclipse.aether.graph.DependencyNode root, List<File> artifacts) {
+        if (root.getArtifact() != null) {
+            artifacts.add(root.getArtifact().getFile());
+        }
+
+        for (org.eclipse.aether.graph.DependencyNode node : root.getChildren()) {
+            addArtifacts(node, artifacts);
+        }
+    }
+
+    private List<Artifact> getNewDependencies(List<Dependency> dependencies, List<Dependency> existingDependencies)
+            throws MojoExecutionException {
+        List<Artifact> updatedArtifacts = new ArrayList<Artifact>();
+        for (Dependency dep : dependencies) {
+            boolean newDependency = true;
+            for (Dependency existingDep : existingDependencies) {
+                if (dep.getArtifactId().equals(existingDep.getArtifactId())) {
+                    newDependency = false;
+                    break;
+                }
+            }
+            if (newDependency) {
+                log.debug("New dependency found: " + dep.getArtifactId());
+                Artifact artifact = getArtifact(dep.getGroupId(), dep.getArtifactId(), dep.getType(), dep.getVersion());
+                updatedArtifacts.add(artifact);
+            }
+        }
+        return updatedArtifacts;
+    }
+
+    private void runTests(String groupId, String artifactId, String phase) throws MojoExecutionException {
+
+        Plugin plugin = project.getPlugin(groupId + ":" + artifactId);
+        if (plugin == null) {
+            plugin = plugin(groupId(groupId), artifactId(artifactId), version("RELEASE"));
+        }
+        Xpp3Dom config = null;
+
+        List<Plugin> buildPlugins = project.getBuildPlugins();
+        for (Plugin p : buildPlugins) {
+            if (p.equals(plugin)) {
+                config = (Xpp3Dom) p.getConfiguration();
+                
+                PluginExecution pe;
+                Map<String, PluginExecution> peMap = p.getExecutionsAsMap();
+                if ((pe = peMap.get("default-" + phase)) != null || (pe = peMap.get(phase)) != null
+                        || (pe = peMap.get("default")) != null) {
+                    Xpp3Dom executionConfig = (Xpp3Dom) pe.getConfiguration();
+                    config = Xpp3Dom.mergeXpp3Dom(executionConfig, config);
+                }
+                break;
+            }
+        }
+
+        if (config == null) {
+            log.debug("Could not find " + artifactId + " configuration for " + phase
+                    + " phase. Creating new configuration.");
+            config = configuration();
```

Comment:
Sorry, where is `configuration()` defined?

Reply:
configuration() is part of MojoExecutor